nsis/src/nsis/file/NsisExecutable.java

```diff
@@ -35,49 +42,117 @@ public NsisExecutable() {
 	 * Creates and initializes a Nsis Executable object
 	 * 
 	 * @param factory      that will be used to create Nsis Executable
-	 * @param byteProvider object that will permit reading bytes from the file.
-	 *                     The lifespan of the byte provider is controlled by
-	 *                     Ghidra.
+	 * @param byteProvider object that will permit reading bytes from the file. The
+	 *                     lifespan of the byte provider is controlled by Ghidra.
 	 * @param layout       object used to load PE executables
 	 * @return The Nsis executable object
 	 * @throws IOException
 	 * @throws InvalidFormatException
 	 */
-	public static NsisExecutable createNsisExecutable(GenericFactory factory,
-			ByteProvider bp, SectionLayout layout)
-			throws IOException, InvalidFormatException {
-		NsisExecutable nsisExecutable = (NsisExecutable) factory
-				.create(NsisExecutable.class);
+	public static NsisExecutable createNsisExecutable(GenericFactory factory, ByteProvider bp,
+			SectionLayout layout) throws IOException, InvalidFormatException {
+		NsisExecutable nsisExecutable = (NsisExecutable) factory.create(NsisExecutable.class);
 		nsisExecutable.initNsisExecutable(factory, bp, layout);
 		return nsisExecutable;
 	}
 
-	private void initNsisExecutable(GenericFactory factory, ByteProvider bp,
-			SectionLayout layout) throws IOException, InvalidFormatException {
-		this.reader = new FactoryBundledWithBinaryReader(factory, bp,
-				/* isLittleEndian= */ true);
+	private void initNsisExecutable(GenericFactory factory, ByteProvider bp, SectionLayout layout)
+			throws IOException, InvalidFormatException {
+		this.reader = new FactoryBundledWithBinaryReader(factory, bp, /* isLittleEndian= */ true);
 		this.headerOffset = findHeaderOffset();
-		initScriptHeader(bp);
+		initScriptHeader();
+		if ((this.scriptHeader.compressedHeaderSize & 0x80000000) != 0) { // Check
+																			// if
+																			// MSB
+																			// is
+																			// set
+			decompressData();
+		}
 	}
 
 	private long findHeaderOffset() throws IOException, InvalidFormatException {
-		for (long headerOffset = 0; headerOffset
-				+ NsisConstants.NSIS_MAGIC.length <= reader
-						.length(); headerOffset++) {
+		for (long headerOffset = 0; headerOffset + NsisConstants.NSIS_SIGINFO.length
+				+ NsisConstants.NSIS_MAGIC.length <= reader.length(); headerOffset++) {
 			byte[] content = reader.readByteArray(headerOffset,
-					NsisConstants.NSIS_MAGIC.length);
-			if (Arrays.equals(NsisConstants.NSIS_MAGIC, content)) {
-				return headerOffset;
+					NsisConstants.NSIS_SIGINFO.length + NsisConstants.NSIS_MAGIC.length);
+			if (Arrays.equals(Bytes.concat(NsisConstants.NSIS_SIGINFO, NsisConstants.NSIS_MAGIC),
```

Comment:
is there Java's equivalent for C++'s std::search to use here?
https://www.cplusplus.com/reference/algorithm/search/

Reply:
I looked for one that would do a search with an InputStream object, but could not find one